### PR TITLE
Fix `undefined` entries in overload table

### PR DESCRIPTION
**Pull request by AlexPerrot**
Fix for issue #3436.

argCount is set on every registered method, in case an overload is registered later.

### CODE REVIEW

**Comment by waywardmonkeys** [comment]
Since this is a pretty obscure issue, would you mind extending the embind test with something to cover this now that you have a way to reproduce it?

**Comment by AlexPerrot** [comment]
Here is a simple test, that indeed fails without the fix and succeeds with it.

**Comment by waywardmonkeys** [comment]
Thanks. I have a better idea now of what's happening here and am going to take a closer look today or tomorrow.

I'll comment back on this PR afterwards.

**Comment by AlexPerrot** [comment]
Hi,
Any news about this ?

**Comment by waywardmonkeys** [comment]
Sorry, I've had visitors from the US for the last few days. I will get to this soon!

**Comment by AlexPerrot** [comment]
Since this is not fixed on current incoming, I had another look today.
The order in which the functions are called doesn't seem wrong to me. When all dependent types are registered, then the function is registered and "forgets" its `argCount` (why wait if the needed types are there ?).

**Comment by AlexPerrot** [comment]
Anyone wanting to merge this, or should I close it ?

**Comment by kripken** [comment]
ping @waywardmonkeys @chadaustin

File: src/embind/embind.js [hunk 1]
```diff
@@ -1947,6 +1947,8 @@ var LibraryEmbind = {
             // Replace the initial unbound-handler-stub function with the appropriate member function, now that all types
             // are resolved. If multiple overloads are registered for this function, the function goes into an overload table.
             if (undefined === proto[methodName].overloadTable) {
+                // Set argCount in case an overload is registered later
+                memberFunction.argCount = argCount - 2;
```
Comment (juj) [inline src/embind/embind.js]: I wonder if this section should also copy `memberFunction.className = classType.name;` over like is done on line 1935 above?

**Comment by juj** [comment]
I'd be comfortable merging this in. One question in the review above. Before merging though, could you rebase this against latest `incoming`, since otherwise the merge comes from a branch very long ago, which makes following history more difficult (sorry about that).

**Comment by chadaustin** [comment]
For the record, I have no objection to this getting merged!  Thanks for your work!  (But if @waywardmonkeys wants to do some further investigation, that's fine with me too.  Sadly, I don't remember a ton of specifics about the workings of these functions.)

File: src/embind/embind.js (hunk 1 above)
Comment (AlexPerrot) [inline src/embind/embind.js]: Probably...
I don't have any use case with inheritance, so I didn't test this.
I'll test it if I have some time.

**Comment by AlexPerrot** [comment]
@juj
> Before merging though, could you rebase this against latest incoming, since otherwise the merge comes from a branch very long ago, which makes following history more difficult (sorry about that).

Actually, I tried to do that a few days ago, but being more than 3000 commits beind, I had to manually resolve conflicts every couple of commits. After half an hour, I had only treated a few hundred commits. Furthermore, I had to look at the current code to know how to solve every conflict. Isn't there an easier way to do that ?

**Comment by AlexPerrot** [comment]
By the way, IIRC, this PR only solves the issue for object methods bindings. Regular functions bindings also need to be fixed the same way, I think.

**Comment by juj** [comment]
You can check out the branch with your changes, then fetch origin up to date, and `git rebase origin/incoming` to take the changes up to latest. There didn't seem to be any conflicts, here's a tree that has the rebase done against latest: https://github.com/juj/emscripten/compare/incoming...alexperrot_incoming

**Comment by AlexPerrot** [comment]
> You can check out the branch with your changes, then fetch origin up to date, and git rebase origin/incoming to take the changes up to latest. There didn't seem to be any conflicts, here's a tree that has the rebase done against latest: juj@incoming...alexperrot_incoming

That's what I did. After the rebase, I get this from git :
```
On branch incoming
Your branch and 'origin/incoming' have diverged,
and have 3182 and 2 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
nothing to commit, working directory clean
```

Conflict happen when using `git pull` as instructed. The first conflict is :
```
rebase in progress; onto 5099f2e
You are currently rebasing branch 'incoming' on '5099f2e'.
  (fix conflicts and then run "git rebase --continue")
  (use "git rebase --skip" to skip this patch)
  (use "git rebase --abort" to check out the original branch)

Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add <file>..." to mark resolution)

	both modified:   AUTHORS
```

For information, `origin/incoming` is the remote branch from my fork, and I did the rebase on `upstream/incoming`, which is `kripken/emscripten`

**Comment by AlexPerrot** [comment]
Ok, `git push --force` was the way to go.
The PR is now in sync with incoming.
Not sure you should merge right away, though, there may be some other things to take care of first (copy `className` and fix this for regular functions, at least).

File: src/embind/embind.js (hunk 1 above)
Comment (AlexPerrot) [inline src/embind/embind.js]: I couldn't find any usefulness to className in my tests.
Either it is different, and the method is replaced, or it is the same, and the code creates an overload table only to immediately replace the only method it contains with the newly registered one. Either way, the method previously registered is lost.

**Comment by AlexPerrot** [comment]
I just pushed a test and a fix for the free functions.
Apart from `className` (see my comment), I think this is done.

**Comment by juj** [comment]
Thanks, pushing this in, great work for having a test to verify this as well.

**Comment by juj** [comment]
Noticed that after merging this, the second late added test case actually fails. Marked this down as https://github.com/kripken/emscripten/issues/4787

**Pull request merged**